### PR TITLE
Add billing tables 

### PR DESCRIPTION
**Pull request by anamarn**
Beforehand, the name of the branch is not representative of the work that has been done in this PR

**TLDR:**

Solves https://github.com/twentyhq/private-issues/issues/192
Add 3 tables BillingCustomer, BillingProduct and BillingPrice  and BillingMeter to core, inspired by the Stripe implementation.  Separates migration, between common and billing on order to not populate the db of the self-hosting instances with unused tables.

**In order to test:**

Run the command:
npx nx typeorm -- migration:run -d src/database/typeorm/core/core.datasource.ts

**Considerations:**

I only put the information we should use right now in the Billing module,  for instance columns like meter or agreggation formula where omitted in the creation of the tables. 
These columns and other ones who fall on the same spectrum will be added as we need them. 

If you want to add more information to the table, I'll leave some utility links down bellow:

- BillingPrices: https://docs.stripe.com/api/prices/object
- BillingCustomer: https://docs.stripe.com/api/customers/object
- BillingProduct:  https://docs.stripe.com/api/products/object

**Next Steps**

Use the Stripe Webhook in order to update the tables accordingly

### CODE REVIEW

**Comment by FelixMalfait** [comment]
This looks great! I'm starting to think that this will really cluter self-hosted instances with tables they don't need. Could you please make is so that the app works without those tables if IS_BILLING_ENABLED is false? I pushed a small commit to help your get in that direction (now depending on the environment variable value it will run a different set of migrations).
Thanks!

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-price.entity.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  @Field(() => SubscriptionInterval, { nullable: true })
+  @Column({
+    type: 'enum',
+    enum: Object.values(SubscriptionInterval),
+    nullable: true,
+  })
+  interval: Stripe.Price.Recurring.Interval;
+
+  @Column({ nullable: false })
+  productId: string;
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-price.entity.ts:60]: Do we need really need this or can we instead build the relationship on BillingPrice. stripeProductId = BillingProduct.stripeId?

**Comment by FelixMalfait** [comment]
Could you please look at the Greptile comments and delete them if not relevant (or address them if they are!)? Thanks!

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-product.entity.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  @Column({ nullable: false })
+  defaultStripePriceId: string;
+
+  @Field(() => BillingAvailablePlanKey, { nullable: false })
+  @Column({
+    type: 'enum',
+    enum: Object.values(BillingAvailablePlanKey),
+    nullable: false,
+  })
+  planKey: BillingAvailablePlanKey;
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-product.entity.ts:51]: Maybe you've discussed that with @charlesBochet  but I guess this comes from metadata column? I would be tempted to use a jsonb metadata column and then cast at the code level to really stick to Stripe as much as we can

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingProduct_plankey_enum" AS ENUM('base-plan', 'pro-plan')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "core"."billingProduct" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "deletedAt" TIMESTAMP WITH TIME ZONE, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "active" boolean NOT NULL, "stripeProductId" character varying NOT NULL, "defaultPriceId" character varying NOT NULL, "defaultStripePriceId" character varying NOT NULL, "planKey" "core"."billingProduct_plankey_enum" NOT NULL, CONSTRAINT "PK_8bb3c7be66db8e05476808b0ca7" PRIMARY KEY ("id"))`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts:14]: missing constraints?

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+export class AddBillingTables1732640291160 implements MigrationInterface {
+  name = 'AddBillingTables1732640291160';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE "core"."billingCustomer" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "deletedAt" TIMESTAMP WITH TIME ZONE, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "stripeCustomerId" character varying NOT NULL, "workspaceId" character varying NOT NULL, CONSTRAINT "IndexOnWorkspaceIdAndStripeCustomerIdUnique" UNIQUE ("workspaceId", "stripeCustomerId"), CONSTRAINT "PK_5fffcd69bf722c297a3d5c3f3bc" PRIMARY KEY ("id"))`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingProduct_plankey_enum" AS ENUM('base-plan', 'pro-plan')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts:11]: planKey (case) but not important

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_usagetype_enum" AS ENUM('metered', 'licensed')`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_interval_enum" AS ENUM('day', 'month', 'week', 'year')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "core"."billingPrice" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "deletedAt" TIMESTAMP WITH TIME ZONE, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "stripePriceId" character varying NOT NULL, "active" boolean NOT NULL, "usageType" "core"."billingPrice_usagetype_enum" NOT NULL, "interval" "core"."billingPrice_interval_enum", "productId" character varying NOT NULL, "stripeProductId" character varying NOT NULL, "billingProductId" uuid, CONSTRAINT "IndexOnProductIdAndStripePriceIdUnique" UNIQUE ("productId", "stripePriceId"), CONSTRAINT "PK_13927aef8d4e68e176a61c33d89" PRIMARY KEY ("id"))`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1732640291160-addBillingTables.ts:23]: I think that these stripeIds are actually unique so unicity constraints do not have to be composed, it won't help

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732710668000-billingTableMigrationTransfer.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    try {
+      await queryRunner.query(
+        `ALTER TABLE "core"."billingSubscription" ALTER COLUMN "status" TYPE text`,
+      );
+      await queryRunner.query(
+        `ALTER TABLE "core"."billingSubscription" ALTER COLUMN "interval" TYPE text`,
+      );
+    } catch (error) {
+      // Column may have already been altered in UseEnumForSubscriptionStatusInterval1719327438923
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1732710668000-billingTableMigrationTransfer.ts:31]: I don't get these comments ?

File: packages/twenty-server/src/database/typeorm/core/core.datasource.ts [hunk 7]
```diff
@@ -18,9 +18,15 @@ export const typeORMCoreModuleOptions: TypeOrmModuleOptions = {
   migrationsRun: false,
   migrationsTableName: '_typeorm_migrations',
   metadataTableName: '_typeorm_generated_columns_and_materialized_views',
-  migrations: [
-    `${isJest ? '' : 'dist/'}src/database/typeorm/core/migrations/*{.ts,.js}`,
-  ],
+  migrations:
+    process.env.IS_BILLING_ENABLED === 'true'
+      ? [
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/core.datasource.ts:23]: I think this works but I'm not sure how it will behave between migrate up and down in case we change IS_BILLING_ENABLED value. Let's say it is not meant to be changed :p

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732710668000-billingTableMigrationTransfer.ts (hunk 6 above)
Comment (charlesBochet) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1732710668000-billingTableMigrationTransfer.ts:31]: got it

Comment (charlesBochet) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1732710668000-billingTableMigrationTransfer.ts:31]: I don't see a better way (we could check table existence before but this looks a bit painful to do)

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1732725228911-modifBillingTable.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscription" DROP CONSTRAINT "FK_a8deacba1a77260507a39c741dc"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingPrice" DROP CONSTRAINT "IndexOnProductIdAndStripePriceIdUnique"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingProduct" DROP COLUMN "defaultPriceId"`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1732725228911-modifBillingTable.ts:14]: let's fix the initial migration instead!

Comment (charlesBochet) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1732725228911-modifBillingTable.ts:14]: same for all updates on product, price here

File: packages/twenty-server/src/engine/core-modules/billing/enums/billing-available-plan-key.enum.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+export enum BillingAvailablePlanKey {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/enums/billing-available-plan-key.enum.ts:1]: BillingPlanKey (not sure what available means here)

File: packages/twenty-server/src/engine/core-modules/billing/services/billing.service.ts [hunk 10]
```diff
@@ -22,7 +22,10 @@ export class BillingService {
     return this.environmentService.get('IS_BILLING_ENABLED');
   }
 
-  async hasWorkspaceActiveSubscriptionOrFreeAccess(workspaceId: string) {
+  async hasWorkspaceActiveSubscriptionOrFreeAccess(
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing.service.ts:25]: function name does not look right anymore

**Review by charlesBochet — COMMENTED** [review]
Good one, left a few comments.
Let's make 100% this does not break existing workspace:
- checkout `main`, reset database with IS_BILLING_ENABLED false
- checkout `feat/add-billing-meter`, migrate

Same with IS_BILLING_ENABLED true

File: packages/twenty-server/src/engine/core-modules/billing/types/billing-product-metadata.type.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+import { BillingAvailablePlanKey } from 'src/engine/core-modules/billing/enums/billing-available-plan-key.enum';
+
+export type BillingProductMetadata = {
+  planKey: BillingAvailablePlanKey;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/types/billing-product-metadata.type.ts:4]: don't we also need to identify if a product is the PlanBaseProduct too (the one bearing the base price)

BillingProductMetadata = {
  planKey: BillingAvailablePlanKey;
  isBasePlan?: boolean;
}

what do you think @FelixMalfait @anamarn ?

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-product.entity.ts (hunk 2 above)
Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/billing/entities/billing-product.entity.ts]: agree, + let's keep some flexibility here as this is flexible in stripe

File: packages/twenty-server/src/engine/core-modules/billing/types/billing-product-metadata.type.ts (hunk 11 above)
Comment (anamarn) [reply packages/twenty-server/src/engine/core-modules/billing/types/billing-product-metadata.type.ts:4]: Yes, you are right this will simplify the queries when accessing the billing tables, the structure looks good to me, i'll implement this in my next commit

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+      `ALTER TABLE "core"."billingSubscriptionItem" DROP CONSTRAINT "IndexOnBillingSubscriptionIdAndStripeSubscriptionItemIdUnique"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" DROP CONSTRAINT "IndexOnBillingSubscriptionIdAndStripeProductIdUnique"`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "core"."billingCustomer" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "deletedAt" TIMESTAMP WITH TIME ZONE, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "workspaceId" uuid NOT NULL, "stripeCustomerId" character varying NOT NULL, CONSTRAINT "UQ_b35a0ef2e2f0d40101dd7f161b9" UNIQUE ("stripeCustomerId"), CONSTRAINT "IndexOnWorkspaceIdAndStripeCustomerIdUnique" UNIQUE ("workspaceId", "stripeCustomerId"), CONSTRAINT "PK_5fffcd69bf722c297a3d5c3f3bc" PRIMARY KEY ("id"))`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingMeter_status_enum" AS ENUM('active', 'inactive')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:22]: ACTIVE, INACTIVE

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingMeter_eventtimewindow_enum" AS ENUM('day', 'hour')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "core"."billingMeter" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "deletedAt" TIMESTAMP WITH TIME ZONE, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), "stripeMeterId" character varying NOT NULL, "displayName" character varying NOT NULL, "eventName" character varying NOT NULL, "status" "core"."billingMeter_status_enum" NOT NULL, "customerMapping" jsonb NOT NULL, "eventTimeWindow" "core"."billingMeter_eventtimewindow_enum", "valueSettings" jsonb NOT NULL, CONSTRAINT "UQ_340c08c4e5dd33cf963cbb133ae" UNIQUE ("stripeMeterId"), CONSTRAINT "PK_0bba5f7d2e3713332a0138ea1b3" PRIMARY KEY ("id"))`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_taxbehavior_enum" AS ENUM('exclusive', 'inclusive', 'unspecified')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:31]: EXCLUSIVE, INCLUSIVE, ...

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_type_enum" AS ENUM('one_time', 'recurring')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:34]: ONE_TIME
RECURRING

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+    );
+    await queryRunner.query(
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:36]: PER_UNIT

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+      `CREATE TYPE "core"."billingPrice_billingscheme_enum" AS ENUM('per_unit', 'tiered')`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_tiersmode_enum" AS ENUM('graduated', 'volume')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:40]: GRADUATED

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+    );
+    await queryRunner.query(
+      `CREATE TYPE "core"."billingPrice_usagetype_enum" AS ENUM('metered', 'licensed')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1733307912282-addNewBillingStripeBasedTables.ts:43]: METERED

File: packages/twenty-server/src/database/typeorm/typeorm.service.ts [hunk 18]
```diff
@@ -36,6 +40,10 @@ export class TypeORMService implements OnModuleInit, OnModuleDestroy {
         FeatureFlagEntity,
         BillingSubscription,
         BillingSubscriptionItem,
+        BillingMeter,
+        BillingCustomer,
+        BillingProduct,
+        BillingPrice,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/typeorm.service.ts:46]: double check

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts [hunk 19]
```diff
@@ -51,9 +62,9 @@ export class BillingSubscriptionItem {
   @Column({ nullable: false })
   stripePriceId: string;
 
-  @Column({ nullable: false })
+  @Column({ nullable: false, unique: true })
   stripeSubscriptionItemId: string;
 
-  @Column({ nullable: false })
+  @Column({ nullable: true })
   quantity: number;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts:69]: number | null

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts [hunk 20]
```diff
@@ -34,7 +35,13 @@ export class BillingSubscriptionItem {
...
 
   @Column({ nullable: false })
-  billingSubscriptionId: string;
+  stripeSubscriptionId: string;
+
+  @Column({ nullable: false, type: 'jsonb', default: {} })
+  metadata: Stripe.Metadata;
+
+  @Column({ nullable: true, type: 'jsonb' })
+  billingThresholds: Stripe.SubscriptionItem.BillingThresholds;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts:44]: | null

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts [hunk 21]
```diff
@@ -34,7 +35,13 @@ export class BillingSubscriptionItem {
   updatedAt: Date;
 
   @Column({ nullable: false })
-  billingSubscriptionId: string;
+  stripeSubscriptionId: string;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription-item.entity.ts:38]: check but this should fail on main (with existing subs)

File: packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription.entity.ts [hunk 22]
```diff
@@ -30,6 +32,10 @@ export class BillingSubscription {
   @PrimaryGeneratedColumn('uuid')
   id: string;
 
+  static transformTimestampToDate(timestamp: number): Date {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/entities/billing-subscription.entity.ts:35]: remove

File: packages/twenty-server/src/engine/core-modules/billing/enums/billing-plan-key.enum.ts [hunk 23]
```diff
@@ -0,0 +1,4 @@
+export enum BillingPlanKey {
+  BASE_PLAN = 'base-plan',
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/enums/billing-plan-key.enum.ts:2]: BASE_PLAN

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/billing/enums/billing-plan-key.enum.ts:2]: if existing we don't touch

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts [hunk 24]
```diff
@@ -54,30 +55,51 @@ export class BillingWebhookService {
...
 
-    const billingSubscription =
-      await this.billingSubscriptionRepository.findOneOrFail({
-        where: { stripeSubscriptionId: data.object.id },
-      });
-
     await this.billingSubscriptionItemRepository.upsert(
       data.object.items.data.map((item) => {
         return {
-          billingSubscriptionId: billingSubscription.id,
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts:72]: revert

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts [hunk 25]
```diff
@@ -54,30 +55,51 @@ export class BillingWebhookService {
...
           stripePriceId: item.price.id,
           stripeSubscriptionItemId: item.id,
           quantity: item.quantity,
+          metadata: item.metadata,
+          billingThresholds: item.billing_thresholds ?? undefined,
         };
       }),
       {
-        conflictPaths: ['billingSubscriptionId', 'stripeProductId'],
+        conflictPaths: ['stripeSubscriptionId', 'stripeProductId'],
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts:102]: revert

**Review by charlesBochet — APPROVED** [review]
